### PR TITLE
Seungheonoh/extramap

### PR DESCRIPTION
**Pull request by SeungheonOh**
replaces #562 due to permission issue.

### CODE REVIEW

File: plutarch-extra/Plutarch/Extra/Map.hs [hunk 1]
```diff
@@ -0,0 +1,324 @@
...
+      forall (s' :: S).
+      Term
+        s'
+        ( (v :--> v)
+            :--> k
+            :--> PBuiltinPair (PAsData k) (PAsData v)
+            :--> PBuiltinPair (PAsData k) (PAsData v)
+        )
+    go = phoistAcyclic $
+      plam $ \f target kv ->
```
Comment (L-as) [inline plutarch-extra/Plutarch/Extra/Map.hs:69]: rename target to key or key to target

File: plutarch-extra/Plutarch/Extra/Map.hs [hunk 2]
```diff
@@ -0,0 +1,324 @@
...
+ have to scan the entire map to find the key in the worst case.
+-}
+padjust ::
+  forall (k :: PType) (v :: PType) (s :: S).
+  (PIsData k, PEq k, PIsData v) =>
+  Term s ((v :--> v) :--> k :--> PMap 'Unsorted k v :--> PMap 'Unsorted k v)
+padjust = phoistAcyclic $
+  plam $ \f key kvs ->
+    pmatch kvs $ \(PMap kvs') ->
+      pcon . PMap $ PList.pmap # (go # f # key) # kvs'
```
Comment (L-as) [inline plutarch-extra/Plutarch/Extra/Map.hs:57]: Do we not have a pmap on maps?

File: plutarch-extra/Plutarch/Extra/Map.hs [hunk 3]
```diff
@@ -0,0 +1,324 @@
...
+
+{- | As 'pkeysEqual', but requires only 'PEq' constraints for the keys, and
+ works for 'Unsorted' 'PMap's. This requires a number of equality comparisons
+ between keys proportional to the product of the lengths of both arguments:
+ that is, this function is quadratic.
+-}
+pkeysEqualUnsorted ::
+  forall (k :: PType) (a :: PType) (b :: PType) (s :: S).
+  (PIsData k, PIsData a, PIsData b) =>
+  Term s (PMap 'Unsorted k a :--> PMap 'Unsorted k b :--> PBool)
```
Comment (L-as) [inline plutarch-extra/Plutarch/Extra/Map.hs:83]: When is this more efficient than just sorting before checking?

File: plutarch-extra/Plutarch/Extra/Map.hs [hunk 4]
```diff
@@ -0,0 +1,324 @@
...
+
+{- | Given a 'Foldable' of key-value pairs, construct an unsorted 'PMap'.
+ Performs linearly with respect to its argument.
+
+ = Note
+
+ If there are duplicate keys in the input, the /last/ key will \'win\' in a
+ lookup.
+-}
+punsortedMapFromFoldable ::
```
Comment (L-as) [inline plutarch-extra/Plutarch/Extra/Map.hs:216]: Isn't this basically an IsList instance? Why not just implement that?

File: plutarch-extra/Plutarch/Extra/Map.hs [hunk 5]
```diff
@@ -0,0 +1,324 @@
...
+
+ If there are duplicate keys, only the /last/ key-value pair will remain in
+ the result.
+-}
+psortedMapFromFoldable ::
+  forall (k :: PType) (v :: PType) (f :: Type -> Type) (s :: S).
+  (Foldable f, POrd k, PIsData k, PIsData v) =>
+  f (Term s k, Term s v) ->
+  Term s (PMap 'Sorted k v)
+psortedMapFromFoldable = foldl' go pempty
```
Comment (L-as) [inline plutarch-extra/Plutarch/Extra/Map.hs:245]: You could also make an instance for this.

File: plutarch-extra/Plutarch/Extra/Map.hs [hunk 6]
```diff
@@ -0,0 +1,324 @@
...
+ 'PNothing', the value is deleted; otherwise, it is modified to the result.
+
+ Performance will be equivalent to a lookup followed by an insert (or delete),
+ as well as the cost of calling the \'updater\'.
+-}
+pupdate ::
+  forall (k :: PType) (v :: PType) (s :: S).
+  (PIsData k, PIsData v, POrd k) =>
+  Term s ((v :--> PMaybe v) :--> k :--> PMap 'Sorted k v :--> PMap 'Sorted k v)
+pupdate = phoistAcyclic $
```
Comment (L-as) [inline plutarch-extra/Plutarch/Extra/Map.hs:296]: This is quite inefficiently implemented.

**Review by L-as — CHANGES_REQUESTED** [review]

**Comment by L-as** [comment]
Thanks!

File: plutarch-extra/Plutarch/Extra/Map.hs (hunk 2 above)
Comment (L-as) [reply plutarch-extra/Plutarch/Extra/Map.hs:57]: Make pmapKeyed?

Comment (SeungheonOh) [reply plutarch-extra/Plutarch/Extra/Map.hs:57]: okay

File: plutarch-extra/Plutarch/Extra/Map.hs (hunk 3 above)
Comment (SeungheonOh) [reply plutarch-extra/Plutarch/Extra/Map.hs:83]: It's more efficient in terms of memories

File: plutarch-extra/Plutarch/Extra/Map.hs (hunk 4 above)
Comment (SeungheonOh) [reply plutarch-extra/Plutarch/Extra/Map.hs:216]: I don't think you can implement `toList` if we were to implement `IsList`

Comment (SeungheonOh) [reply plutarch-extra/Plutarch/Extra/Map.hs:216]: at list without `plift`s which we don't want.

File: plutarch-extra/Plutarch/Extra/Map.hs (hunk 3 above)
Comment (L-as) [reply plutarch-extra/Plutarch/Extra/Map.hs:82]: You should note that

File: plutarch-extra/Plutarch/Extra/Map.hs (hunk 4 above)
Comment (L-as) [reply plutarch-extra/Plutarch/Extra/Map.hs:215]: I'm not sure what you mean.

Comment (SeungheonOh) [reply plutarch-extra/Plutarch/Extra/Map.hs:215]: https://github.com/Liqwid-Labs/plutarch-quickcheck/blob/34f1157dd5d68d289beda59fc5c6f68da42d42c9/src/Plutarch/Test/QuickCheck/Instances.hs#L829

I can't think of a better way here. And I'm sure this isn't what we want.

Comment (kozross) [reply plutarch-extra/Plutarch/Extra/Map.hs:215]: Yeah, the `toList` direction would be _horrid_.

Comment (L-as) [reply plutarch-extra/Plutarch/Extra/Map.hs:215]:
```haskell
instance IsList (Term s (PMap 'Unsorted k v)) where
  type Item (Term s (PMap 'Unsorted k v)) = [(Term s k, Term s v)]
  fromList = punsortedMapFromFoldable

instance IsList (Term s (PMap 'Sorted k v)) where
  type Item (Term s (PMap 'Unsorted k v)) = [(Term s k, Term s v)]
  fromList = psortedMapFromFoldable
```

Comment (SeungheonOh) [reply plutarch-extra/Plutarch/Extra/Map.hs:215]: wait we can make `IsList` with no `toList`?

File: plutarch-extra/Plutarch/Extra/Map.hs [hunk 7]
```diff
@@ -250,6 +254,16 @@ psortedMapFromFoldable = foldl' go pempty
       Term s' (PMap 'Sorted k v)
     go acc (key, val) = pinsert # key # val # acc
 
+instance (PIsData k, PIsData v, POrd k) => IsList (Term s (PMap 'Unsorted k v)) where
```
Comment (L-as) [inline plutarch-extra/Plutarch/Extra/Map.hs:257]: Orphan instances are very bad. Just move the relevant functions into the defining module.

Comment (L-as) [reply plutarch-extra/Plutarch/Extra/Map.hs:257]: to the module that defines PMap

Comment (SeungheonOh) [reply plutarch-extra/Plutarch/Extra/Map.hs]: Ah, okay. Thank you.

**Review by L-as — APPROVED** [review]
LGTM

**Pull request merged**